db/warehouse/migrate/20241010005805_prune_warehouse_indexes_phase_1.rb

```diff
@@ -0,0 +1,542 @@
+    # done carefully
+    raise ActiveRecord::IrreversibleMigration
+  end
+
+  protected
+
+  def check_index_table(index_name, table_name)
+    sql = <<~SQL
+      SELECT t.relname
+      FROM pg_index i
+      JOIN pg_class t ON t.oid = i.indrelid
+      JOIN pg_class idx ON idx.oid = i.indexrelid
+      WHERE idx.relname = $1
+    SQL
+    result = GrdaWarehouseBase.connection.select_value(sql, 'SQL', [index_name])
+    raise "expected index #{index_name} to be on #{table_name} but was #{result}" unless result == table_name
+  end
+
+  # it seems that the index names aren't consistent between prod and dev. Attempt to remove the index by name only
+  def remove_index_safely(table:, index_name:, columns:)
+    if index_exists?(table, index_name)
+      check_index_table(index_name, table) # additional safety check
+      remove_index(table, name: index_name)
+    # we could try and remove equivalent index but there's less certainty that is is unused if the name doesn't match
+    # elsif Rails.env.development? && index_exists?(table, columns)
+    #   remove_index(table, column: columns)
+    else
+      msg = "Index not found for table: #{table}, columns: #{columns.join(', ')}"
+      Rails.logger.error(msg)
+    end
+  end
+
+  ## dry-run
+  # def remove_index(table, columns: nil, name: nil)
+  #   puts "would remove: #{table}, columns: #{columns.inspect}, name: #{name}"
+  # end
+
+  def parse_definition(definition)
+    # Parse the CREATE INDEX statement
+    match = definition.match(/CREATE INDEX "?(?<index_name>[^"\s]+)"? ON (?:public\.)?(?<table>[^\s]+).*\((?<columns>.*)\)/)
+
+    raise "Failed to parse index definition: #{definition}" unless match
+
+    table = match[:table].gsub('"', '')
+    index_name = match[:index_name]
+    columns = match[:columns].split(',').map { |col| col.strip.gsub('"', '') }
+
+    [table, index_name, columns]
+  end
+
+  # index definitions that seem to be unused in production.
+  #   1) original sql query
+  #     -- original query
+  #     SELECT
+  #       pg_get_indexdef(indexrelid)
+  #       FROM
+  #         pg_stat_user_indexes s
+  #         JOIN pg_index i USING (indexrelid)
+  #         JOIN pg_class c ON c.oid = i.indrelid
+  #       WHERE
+  #         schemaname = 'public'
+  #         AND s.idx_scan = 0     -- has never been scanned
+  #         AND 0 <>ALL (i.indkey) -- no index column is an expression
+  #         AND NOT i.indisunique  -- is not a UNIQUE index
+  #         AND NOT EXISTS (       -- does not enforce a constraint
+  #           SELECT
+  #             1
+  #           FROM
+  #             pg_catalog.pg_constraint con
+  #           WHERE
+  #             con.conindid = s.indexrelid
+  #         );
+  #  2) using our custom app inspector data, restrict the above set to only those that are reported as unused across a
+  #     a reasonable sample of our installations. See Dba::UnusedWarehouseIndexMigrationHelper
+  INDEX_TEXT = <<~TXT.freeze
```

Comment:
I think this should be moved into `Dba::UnusedWarehouseIndexMigrationHelper` since future installations may need these removed as well.  I guess they won't be in the structure file so new installations won't get them, so maybe that's ok, but once we collapse migrations again, we'll lose this list.

Reply:
Sure, I can move them.
I assume we want them to fade out eventually but if you think these indexes might get recreated somehow we can retain them for reference. Ideally we would track down whatever is creating these indexes and fix it

I'll add a todo-or-die to UnusedWarehouseIndexMigrationHelper so we remember to take out the garbage.